File: src/main/java/org/jitsi/jicofo/bridge/Bridge.java
/*
 * Jicofo, the Jitsi Conference Focus.
 *
 * Copyright @ 2015-Present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jitsi.jicofo.bridge;

import edu.umd.cs.findbugs.annotations.*;
import org.jitsi.jicofo.xmpp.*;
import org.jitsi.utils.*;
import org.jitsi.utils.stats.*;
import org.jitsi.xmpp.extensions.colibri.*;
import org.jxmpp.jid.*;

import java.time.*;

import static org.jitsi.xmpp.extensions.colibri.ColibriStatsExtension.*;
import static org.jitsi.jicofo.bridge.BridgeConfig.config;

/**
 * Represents a jitsi-videobridge instance, reachable at a certain JID, which
 * can be used by jicofo for hosting conferences. Contains the state related
 * to the jitsi-videobridge instance, such as numbers of channels and streams,
 * the region in which the instance resides, etc.
 *
 * TODO fix comparator (should not be reflexive unless the objects are the same?)
 * @author Pawel Domas
 * @author Boris Grozev
 */
@SuppressFBWarnings("EQ_COMPARETO_USE_OBJECT_EQUALS")
public class Bridge
    implements Comparable<Bridge>
{
    /**
     * How long the "failed" state should be sticky for. Once a {@link Bridge} goes in a non-operational state (via
     * {@link #setIsOperational(boolean)}) it will be considered non-operational for at least this amount of time.
     * See the tests for example behavior.
     */
    private static final Duration failureResetThreshold = config.failureResetThreshold();

    /**
     * The XMPP address of the bridge.
     */
    @NonNull private final Jid jid;

    /**
     * Keep track of the recently added endpoints.
     */
    private final RateTracker newEndpointsRate
            = new RateTracker(config.participantRampupInterval(), Duration.ofMillis(100));

    /**
     * The last reported packet rate in packets per second.
     */
    private int lastReportedPacketRatePps = 0;

    /**
     * The last report stress level
     */
    private double lastReportedStressLevel = 0.0;

    /**
     * For older bridges which don't support reporting their stress level we'll fall back
     * to calculating the stress manually via the packet rate.
     */
    private boolean usePacketRateStatForStress = true;

    /**
     * Holds bridge version (if known - not all bridge version are capable of
     * reporting it).
     */
    private String version = null;

    /**
     * Holds bridge release ID, or null if not known.
     */
    private String releaseId = null;

    /**
     * Whether the bridge supports colibri2.
     */
    private boolean colibri2 = false;

    /**
     * Stores the {@code operational} status of the bridge, which is
     * {@code true} if the bridge has been successfully used by the focus to
     * allocate channels. It is reset to {@code false} when the focus fails
     * to allocate channels, but it gets another chance when all currently
     * working bridges go down and might eventually get elevated back to
     * {@code true}.
     */
    private volatile boolean isOperational = true;

    /**
     * Start out with the configured value, update if the bridge reports a value.
     */
    private double averageParticipantStress = config.averageParticipantStress();

    /**
     * Stores a boolean that indicates whether the bridge is in graceful shutdown mode.
     */
    private boolean shutdownInProgress = false /* we assume it is not shutting down */;

    /**
     * Stores a boolean that indicates whether the bridge is in drain mode.
     */
    private boolean draining = true; /* Default to true to prevent unwanted selection before reading actual state */

    /**
     * The time when this instance has failed.
     *
     * Use {@code null} to represent "never" because calculating the duration from {@link Instant#MIN} is slow.
     */
    @Nullable
    private Instant failureInstant = null;

    private String region = null;
    private String relayId = null;

    @NonNull
    private final Clock clock;

    Bridge(@NonNull Jid jid, @NonNull Clock clock)
    {
        this.jid = jid;
        this.clock = clock;
    }

    Bridge(@NonNull Jid jid)
    {
        this(jid, Clock.systemUTC());
    }

    /**
     * Notifies this instance that a new {@link ColibriStatsExtension} was
     * received for this instance.
     * @param stats the {@link ColibriStatsExtension} instance which was
     * received.
     */
    public void setStats(ColibriStatsExtension stats)
    {
        if (stats == null)
        {
            return;
        }

        Double stressLevel = UtilKt.getDouble(stats, "stress_level");
        if (stressLevel != null)
        {
            lastReportedStressLevel = stressLevel;
            usePacketRateStatForStress = false;
        }
        Double averageParticipantStress = UtilKt.getDouble(stats, "average_participant_stress");
        if (averageParticipantStress != null)
        {
            this.averageParticipantStress = averageParticipantStress;
        }

        Integer packetRateDown = null;
        Integer packetRateUp = null;
        try
        {
            packetRateDown = stats.getValueAsInt(PACKET_RATE_DOWNLOAD);
            packetRateUp = stats.getValueAsInt(PACKET_RATE_UPLOAD);
        }
        catch (NumberFormatException ignored)
        {
        }

        if (packetRateDown != null && packetRateUp != null)
        {
            lastReportedPacketRatePps = packetRateDown + packetRateUp;
        }

        if (Boolean.parseBoolean(stats.getValueAsString(SHUTDOWN_IN_PROGRESS)))
        {
            shutdownInProgress = true;
        }

        String drainStr = stats.getValueAsString(DRAIN);
        if (drainStr != null)
        {
            draining = Boolean.parseBoolean(drainStr);
        }

        if (Boolean.parseBoolean(stats.getValueAsString("colibri2")))
        {
            colibri2 = true;
        }

        String newVersion = stats.getValueAsString(VERSION);
        if (newVersion != null)
        {
            version = newVersion;
        }

        String newReleaseId = stats.getValueAsString(RELEASE);
        if (newReleaseId != null)
        {
            releaseId = newReleaseId;
        }

        String region = stats.getValueAsString(REGION);
        if (region != null)
        {
            this.region = region;
        }

        String relayId = stats.getValueAsString(RELAY_ID);
        if (relayId != null)
        {
            this.relayId = relayId;
        }
    }

    public boolean supportsColibri2()
    {
        return colibri2;
    }

    /**
     * @return the relay ID advertised by the bridge, or {@code null} if
     * none was advertised.
     */
    public String getRelayId()
    {
        return relayId;
    }

    public void setIsOperational(boolean isOperational)
    {
        this.isOperational = isOperational;

        if (!isOperational)
        {
            // Remember when the bridge has last failed
            failureInstant = clock.instant();
        }
    }

    public boolean isOperational()
    {
        // To filter out intermittent failures, do not return operational
        // until past the reset threshold since the last failure.
        if (failureInstant != null &&
                Duration.between(failureInstant, clock.instant()).compareTo(failureResetThreshold) < 0)
        {
            return false;
        }

        return isOperational;
    }

    /**
     * Returns a negative number if this instance is more able to serve conferences than o. For details see
     * {@link #compare(Bridge, Bridge)}.
     *
     * @param o the other bridge instance
     *
     * @return a negative number if this instance is more able to serve conferences than o
     */
    @Override
    public int compareTo(Bridge o)
    {
        return compare(this, o);
    }

    /**
     * Returns a negative number if b1 is more able to serve conferences than b2. The computation is based on the
     * following three comparisons
     *
     * operating bridges < non operating bridges
     * not in graceful shutdown mode < bridges in graceful shutdown mode
     * lower stress < higher stress
     *
     * @param b1 the 1st bridge instance
     * @param b2 the 2nd bridge instance
     *
     * @return a negative number if b1 is more able to serve conferences than b2
     */
    public static int compare(Bridge b1, Bridge b2)
    {
        int myPriority = getPriority(b1);
        int otherPriority = getPriority(b2);

        if (myPriority != otherPriority)
        {
            return myPriority - otherPriority;
        }

        return Double.compare(b1.getStress(), b2.getStress());
    }

    private static int getPriority(Bridge b)
    {
        return b.isOperational() ? (b.isInGracefulShutdown() ? 2 : 1) : 3;
    }

    /**
     * Notifies this {@link Bridge} that it was used for a new endpoint.
     */
    public void endpointAdded()
    {
        newEndpointsRate.update(1);
    }

    /**
     * Returns the net number of video channels recently allocated or removed
     * from this bridge.
     */
    private long getRecentlyAddedEndpointCount()
    {
        return newEndpointsRate.getAccumulatedCount();
    }

    public Jid getJid()
    {
        return jid;
    }

    /**
     * Get the version of this bridge (with embedded release ID, if available).
     * @return The version, or null if not known.
     */
    public String getVersion()
    {
        if (version != null && releaseId != null)
            return version + "-" + releaseId;
        else
            return version;
    }

    /**
     * @return the region of this {@link Bridge}.
     */
    public String getRegion()
    {
        return region;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString()
    {
        return String.format(
                "Bridge[jid=%s, relayId=%s, region=%s, stress=%.2f]",
                     jid.toString(),
                     getRelayId(),
                     getRegion(),
                     getStress());
    }

    /**
     * Gets the "stress" of the bridge, represented as a double between 0 and 1 (though technically the value
     * can exceed 1).
     * @return this bridge's stress level
     */
    public double getStress()
    {
        if (usePacketRateStatForStress)
        {
            return getStressFromPacketRate();
        }
        // While a stress of 1 indicates a bridge is fully loaded, we allow
        // larger values to keep sorting correctly.
        return (lastReportedStressLevel + Math.max(0, getRecentlyAddedEndpointCount()) * averageParticipantStress);
    }

    /**
     * Returns the "stress" of the bridge. The stress is computed based on the
     * total packet rate reported by the bridge and the video stream diff
     * estimation since the last update from the bridge. Note that this is techincally
     * deprecated and only exists for backwards compatibility with bridges who don't
     * yet support reporting their stress level directly.
     *
     * @return the sum of the last total reported packet rate (in pps) and an
     * estimation of the packet rate of the streams that we estimate that the bridge
     * hasn't reported to Jicofo yet. The estimation is the product of the
     * number of unreported streams and a constant C (which we set to 500 pps).
     */
    private double getStressFromPacketRate()
    {
        // While a stress of 1 indicates a bridge is fully loaded, we allow
        // larger values to keep sorting correctly.
        return (lastReportedPacketRatePps
            + Math.max(0, getRecentlyAddedEndpointCount()) * config.averageParticipantPacketRatePps())
            / (double) config.maxBridgePacketRatePps();
    }

    /**
     * @return true if the stress of the bridge is greater-than-or-equal to the threshold.
     */
    public boolean isOverloaded()
    {
        return getStress() >= config.stressThreshold();
    }

    public double getLastReportedStressLevel()
    {
        return lastReportedStressLevel;
    }

    public boolean isInGracefulShutdown()
    {
        return shutdownInProgress;
    }

    /**
     * @return true if the bridge is currently in drain mode
     */
    public boolean isDraining()
    {
        return draining;
    }

    @NonNull
    public OrderedJsonObject getDebugState()
    {
        OrderedJsonObject o = new OrderedJsonObject();
        o.put("version", String.valueOf(version));
        o.put("release", String.valueOf(releaseId));
        o.put("stress", getStress());
        o.put("operational", isOperational());
        o.put("packet_rate", lastReportedPacketRatePps);
        o.put("region", String.valueOf(region));
        o.put("drain", isDraining());
        o.put("graceful-shutdown", isInGracefulShutdown());
        o.put("overloaded", isOverloaded());
        o.put("relay-id", String.valueOf(relayId));

        return o;
    }
}

File: jicofo-selector/src/main/kotlin/org/jitsi/jicofo/bridge/Bridge.kt
/*
 * Jicofo, the Jitsi Conference Focus.
 *
 * Copyright @ 2015-Present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jitsi.jicofo.bridge

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import org.jitsi.utils.OrderedJsonObject
import org.jitsi.utils.logging2.Logger
import org.jitsi.utils.logging2.LoggerImpl
import org.jitsi.utils.stats.RateTracker
import org.jitsi.xmpp.extensions.colibri.ColibriStatsExtension
import org.jxmpp.jid.Jid
import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.max
import org.jitsi.jicofo.bridge.BridgeConfig.Companion.config as config

/**
 * Represents a jitsi-videobridge instance, reachable at a certain JID, which
 * can be used by jicofo for hosting conferences. Contains the state related
 * to the jitsi-videobridge instance, such as numbers of channels and streams,
 * the region in which the instance resides, etc.
 *
 * TODO fix comparator (should not be reflexive unless the objects are the same?)
 * @author Pawel Domas
 * @author Boris Grozev
 */
@SuppressFBWarnings("EQ_COMPARETO_USE_OBJECT_EQUALS")
class Bridge @JvmOverloads internal constructor(
    /**
     * The XMPP address of the bridge.
     */
    val jid: Jid,
    private val clock: Clock = Clock.systemUTC()
) : Comparable<Bridge> {

    /** Keep track of the recently added endpoints. */
    private val newEndpointsRate = RateTracker(
        config.participantRampupInterval,
        Duration.ofMillis(100),
        clock
    )

    private val endpointRestartRequestRate = RateTracker(
        config.iceFailureDetection.interval,
        Duration.ofSeconds(1),
        clock
    )

    /** Number of endpoints currently allocated on this bridge by this jicofo instance. */
    val endpoints = AtomicInteger(0)

    /** The last report stress level */
    var lastReportedStressLevel = 0.0
        private set

    /** Holds bridge version (if known - not all bridge version are capable of reporting it). */
    private var version: String? = null

    /** Whether the last received presence indicated the bridge is healthy. */
    var isHealthy = true
        private set

    /** Holds bridge release ID, or null if not known. */
    private var releaseId: String? = null

    /**
     * Stores the `operational` status of the bridge, which is
     * `true` if the bridge has been successfully used by the focus to
     * allocate channels. It is reset to `false` when the focus fails
     * to allocate channels, but it gets another chance when all currently
     * working bridges go down and might eventually get elevated back to
     * `true`.
     */
    @Volatile
    var isOperational = true
        get() =
            // To filter out intermittent failures, do not return operational
            // until past the reset threshold since the last failure.
            if (failureInstant != null &&
                Duration.between(failureInstant, clock.instant()).compareTo(config.failureResetThreshold) < 0
            ) {
                false
            } else {
                field
            }
        set(isOperational) {
            field = isOperational
            if (!isOperational) {
                // Remember when the bridge has last failed
                failureInstant = clock.instant()
            }
        }

    /** Start out with the configured value, update if the bridge reports a value. */
    private var averageParticipantStress = config.averageParticipantStress

    /** Stores a boolean that indicates whether the bridge is in graceful shutdown mode. */
    var isInGracefulShutdown = false // we assume it is not shutting down

    /** Whether the bridge is in SHUTTING_DOWN mode. */
    var isShuttingDown = false
        private set

    /**
     * Stores a boolean that indicates whether the bridge is in drain mode.
     */
    var isDraining = true // Default to true to prevent unwanted selection before reading actual state
        private set

    /**
     * The time when this instance has failed.
     *
     * Use `null` to represent "never" because calculating the duration from [Instant.MIN] is slow.
     */
    private var failureInstant: Instant? = null

    /** @return the region of this [Bridge]. */
    var region: String? = null
        private set

    /** @return the relay ID advertised by the bridge, or `null` if none was advertised. */
    var relayId: String? = null
        private set

    /**
     * If this [Bridge] has been removed from the list of bridges. Once removed, the metrics specific to this instance
     *  are cleared and no longer emitted. If the bridge re-connects, a new [Bridge] instance will be created.
     */
    val removed = AtomicBoolean(false)

    /**
     * The last instant at which we detected, based on restart requests from endpoints, that this bridge is failing ICE
     */
    private var lastIceFailed = Instant.MIN
    private val failingIce: Boolean
        get() = Duration.between(lastIceFailed, clock.instant()) < config.iceFailureDetection.timeout

    private val logger: Logger = LoggerImpl(Bridge::class.java.name)

    init {
        logger.addContext("jid", jid.toString())
    }

    private var lastPresenceReceived = Instant.MIN

    val timeSinceLastPresence: Duration
        get() = Duration.between(lastPresenceReceived, clock.instant())

    /**
     * Notifies this instance that a new [ColibriStatsExtension] was
     * received for this instance.
     * @param stats the [ColibriStatsExtension] instance which was
     * received.
     */
    fun setStats(stats: ColibriStatsExtension?) {
        if (stats == null) {
            return
        }
        lastPresenceReceived = clock.instant()
        val stressLevel = stats.getDouble("stress_level")
        if (stressLevel != null) {
            lastReportedStressLevel = stressLevel
        }
        val averageParticipantStress = stats.getDouble("average_participant_stress")
        if (averageParticipantStress != null) {
            this.averageParticipantStress = averageParticipantStress
        }
        if (java.lang.Boolean.parseBoolean(
                stats.getValueAsString(
                    ColibriStatsExtension.SHUTDOWN_IN_PROGRESS
                )
            )
        ) {
            isInGracefulShutdown = true
        }
        if (java.lang.Boolean.parseBoolean(stats.getValueAsString("shutting_down"))) {
            isShuttingDown = true
        }
        val drainStr = stats.getValueAsString(ColibriStatsExtension.DRAIN)
        if (drainStr != null) {
            isDraining = java.lang.Boolean.parseBoolean(drainStr)
        }
        val newVersion = stats.getValueAsString(ColibriStatsExtension.VERSION)
        if (newVersion != null) {
            version = newVersion
        }
        val newReleaseId = stats.getValueAsString(ColibriStatsExtension.RELEASE)
        if (newReleaseId != null) {
            releaseId = newReleaseId
        }
        val region = stats.getValueAsString(ColibriStatsExtension.REGION)
        if (region != null) {
            this.region = region
        }
        val relayId = stats.getValueAsString(ColibriStatsExtension.RELAY_ID)
        if (relayId != null) {
            this.relayId = relayId
        }
        val healthy = stats.getValueAsString("healthy")
        if (healthy != null) {
            isHealthy = java.lang.Boolean.parseBoolean(healthy)
        } else if (config.usePresenceForHealth) {
            logger.warn(
                "Presence-based health checks are enabled, but presence did not include health status. Health " +
                    "checks for this bridge are effectively disabled."
            )
        }
    }

    /**
     * Returns a negative number if this instance is more able to serve conferences than o. For details see
     * [.compare].
     *
     * @param other the other bridge instance
     *
     * @return a negative number if this instance is more able to serve conferences than o
     */
    override fun compareTo(other: Bridge): Int {
        return compare(this, other)
    }

    /** Notifies this [Bridge] that it was used for a new endpoint. */
    fun endpointAdded() {
        newEndpointsRate.update(1)
        endpoints.incrementAndGet()
        if (!removed.get()) {
            BridgeMetrics.endpoints.set(endpoints.get().toLong(), listOf(jid.resourceOrEmpty.toString()))
        }
    }

    /** Updates the "endpoints moved" metric for this bridge. */
    fun endpointsMoved(count: Long) {
        BridgeMetrics.endpointsMoved.add(count, listOf(jid.resourceOrEmpty.toString()))
    }
    fun endpointRemoved() = endpointsRemoved(1)
    fun endpointsRemoved(count: Int) {
        endpoints.addAndGet(-count)
        if (!removed.get()) {
            BridgeMetrics.endpoints.set(endpoints.get().toLong(), listOf(jid.resourceOrEmpty.toString()))
        }
        if (endpoints.get() < 0) {
            logger.error("Removed more endpoints than were allocated. Resetting to 0.", Throwable())
            endpoints.set(0)
        }
    }
    internal fun markRemoved() {
        if (removed.compareAndSet(false, true)) {
            BridgeMetrics.restartRequestsMetric.remove(listOf(jid.resourceOrEmpty.toString()))
            BridgeMetrics.endpoints.remove(listOf(jid.resourceOrEmpty.toString()))
            BridgeMetrics.failingIce.remove(listOf(jid.resourceOrEmpty.toString()))
            BridgeMetrics.endpointsMoved.remove(listOf(jid.resourceOrEmpty.toString()))
        }
    }
    internal fun updateMetrics() {
        if (!removed.get()) {
            BridgeMetrics.failingIce.set(failingIce, listOf(jid.resourceOrEmpty.toString()))
        }
    }

    fun endpointRequestedRestart() {
        endpointRestartRequestRate.update(1)
        if (!removed.get()) {
            BridgeMetrics.restartRequestsMetric.inc(listOf(jid.resourceOrEmpty.toString()))
        }

        if (config.iceFailureDetection.enabled) {
            val restartCount = endpointRestartRequestRate.getAccumulatedCount()
            val endpoints = endpoints.get()
            if (endpoints >= config.iceFailureDetection.minEndpoints &&
                restartCount > endpoints * config.iceFailureDetection.threshold
            ) {
                // Reset the timeout regardless of the previous state, but only log if the state changed.
                if (!failingIce) {
                    logger.info("Detected an ICE failing state.")
                }
                lastIceFailed = clock.instant()
            }
        }
    }

    /** Returns the net number of video channels recently allocated or removed from this bridge. */
    private val recentlyAddedEndpointCount: Long
        get() = newEndpointsRate.getAccumulatedCount()

    /** The version of this bridge (with embedded release ID, if available). */
    val fullVersion: String?
        get() = if (version != null && releaseId != null) "$version-$releaseId" else version

    override fun toString(): String {
        return String.format(
            "Bridge[jid=%s, version=%s, relayId=%s, region=%s, correctedStress=%.2f]",
            jid.toString(),
            fullVersion,
            relayId,
            region,
            correctedStress
        )
    }

    /**
     * Gets the "stress" of the bridge, represented as a double between 0 and 1 (though technically the value
     * can exceed 1).
     * @return this bridge's stress level
     */
    val correctedStress: Double
        get() {
            // Correct for recently added endpoints.
            // While a stress of 1 indicates a bridge is fully loaded, we allow larger values to keep sorting correctly.
            val s = lastReportedStressLevel + recentlyAddedEndpointCount.coerceAtLeast(0) * averageParticipantStress

            // Correct for failing ICE.
            return if (failingIce) max(s, config.stressThreshold + 0.01) else s
        }

    /** @return true if the stress of the bridge is greater-than-or-equal to the threshold. */
    val isOverloaded: Boolean
        get() = correctedStress >= config.stressThreshold

    val debugState: OrderedJsonObject
        get() = OrderedJsonObject().apply {
            this["corrected-stress"] = correctedStress
            this["drain"] = isDraining
            this["endpoints"] = endpoints.get()
            this["endpoint-restart-requests"] = endpointRestartRequestRate.getAccumulatedCount()
            this["failing-ice"] = failingIce
            this["graceful-shutdown"] = isInGracefulShutdown
            this["healthy"] = isHealthy
            this["operational"] = isOperational
            this["overloaded"] = isOverloaded
            this["region"] = region.toString()
            this["relay-id"] = relayId.toString()
            this["release"] = releaseId.toString()
            this["shutting-down"] = isShuttingDown
            this["stress"] = lastReportedStressLevel
            this["version"] = version.toString()
        }

    companion object {

        /**
         * Returns a negative number if b1 is more able to serve conferences than b2. The computation is based on the
         * following three comparisons
         *
         * operating bridges < non operating bridges
         * not in graceful shutdown mode < bridges in graceful shutdown mode
         * lower stress < higher stress
         *
         * @param b1 the 1st bridge instance
         * @param b2 the 2nd bridge instance
         *
         * @return a negative number if b1 is more able to serve conferences than b2
         */
        fun compare(b1: Bridge, b2: Bridge): Int {
            val myPriority = getPriority(b1)
            val otherPriority = getPriority(b2)
            return if (myPriority != otherPriority) {
                myPriority - otherPriority
            } else {
                b1.correctedStress.compareTo(b2.correctedStress)
            }
        }

        private fun getPriority(b: Bridge): Int {
            return if (b.isOperational) {
                if (b.isInGracefulShutdown) 2 else 1
            } else {
                3
            }
        }
    }
}
